feat: EDM4HEPSchema and Newstyle FCCSchema

prayagyadav opened the pull request:
@lgray @davidlange6 @gomber

Here is a clean draft for the EDM4HEPSchema (edm4hep1) and FCCSchema based on the same. I have not yet managed to add many comments and descriptions, but I plan to add them eventually.

**Workings:**

- The EDM4HEPSchema reads the `edm4hep.yaml` file from the assets directory. I felt this was necessary to add maximum functionality to the schema. Reading the specifications of all the 'components' and 'datatypes' from the yaml file helps to identify the 'members' (example, `energy` is a member of `edm4hep::ReconstructedParticle` datatype) and which members correspond to the various types of cross-branch relations in EDM4HEP: vector members, OneToOneRelations, OneToManyRelations and Links.

- The Schema fetches the comments in the `edm4hep.yaml` file and assigns them as docstrings to the relevant branches.

- The EDM4HEPSchema supports all these relations (With Links needing some manual boilerplate code from the user).

- The version of the `edm4hep.yaml` file used is [here](https://github.com/key4hep/EDM4hep/blob/c6bc8f4428aaba54c621de3ae4c4b1c791018aca/edm4hep.yaml). Please note that the way Links are represented in EDM4HEP has changed in the latest commit. @tmadlener can comment more on this. In any case, it seems necessary to find a way to track the changes from edm4hep.yaml, so that the COFFEA EDM4HEPSchema does not become obsolete after a few version changes.

**Link to example Notebooks:**
- [Here is the link to a notebook that describes the features of the EDM4HEPSchema.](https://github.com/prayagyadav/coffea/blob/master/dev/EDM4HEP-Showcase.ipynb)
- [Here is the link to a notebook that describes the features of the FCCSchema that is based on the EDM4HEPSchema.](https://github.com/prayagyadav/coffea/blob/master/dev/FCC-EDM4HEP-Showcase.ipynb)

**Tests:**
- The test sample for EDM4HEP was obtained from [here](https://github.com/key4hep/EDM4hep/blob/main/scripts/createEDM4hepFile.py) (which, I believe, was pointed out by @tmadlener).
- The test sample for the Newstyle (edm4hep1) FCC was provided by @davidlange6
- Both the schema pass the basic tests (the tests are similar to the ones that I wrote for the oldstyle (pre-edm4hep1) FCCSchema)

**Other comments:**
- I understand that this commit includes a lot of code, and its purpose may not be immediately clear, but I can provide a detailed explanation for specific snippets if needed.
- While I have tried to add most of the functionality to the schema, I cannot figure out how to add the `ExtraCode` sections mentioned in `edm4hep.yaml`. They appear to be declarations for C++ methods specific to certain collections.

prayagyadav committed:
EDM4HEPSchema and Newstyle FCCSchema

lgray commented on tests/samples/p8_ee_WW_ecm240_edm4hep.root:
is there any way this file can be made smaller? I'd prefer not to take a 12MB file into the repository.

prayagyadav replied:
Sure, made smaller : 12 M --> 186 K

lgray commented on tests/test_nanoevents_edm4hep.py:
```diff
@@ -0,0 +1,87 @@
+    assert field in delayed_fields
+
+
+def test_MC_daughters(delayed_events):
```
these tests should also check correctness!

lgray commented on tests/test_nanoevents_fcc_edm4hep1.py:
```diff
@@ -0,0 +1,138 @@
+    assert field in delayed_fields
+
+
+def test_MC_daughters(delayed_events):
```
these tests should also check correctness!

lgray commented on tests/test_nanoevents_fcc_edm4hep1.py:
```diff
@@ -0,0 +1,138 @@
+#
+
+
+def test_KaonParent_to_PionDaughters_Loop(eager_events):
```
ah ok, I guess correctness is checked here in the end.

Perhaps some more basic checks in the prior tests would still be useful?

prayagyadav replied:
Yeah, test_KaonParent_to_PionDaughters_Loop attempts to check the correctness of Parents and Daughters relations. Sure, I can add more basic tests

lgray committed:
Merge branch 'master' into edm4hep-schema

lgray commented on src/coffea/nanoevents/transforms.py:
```diff
@@ -282,20 +281,494 @@ def index_range(stack):
+        (begin[:, :, numpy.newaxis], end[:, :, numpy.newaxis]), axis=2
+    )
+
+    @numba.njit
```
Having this be a local function seems like a possible source of slowdown / confusion, please move this into the module namespace!

lgray commented on src/coffea/nanoevents/transforms.py:
```diff
@@ -282,20 +281,494 @@ def index_range(stack):
+
+
+def get_array_from_indices(indices, target):
+    @numba.jit
```
same thing here - no reason to have this locally scoped

lgray commented on src/coffea/nanoevents/transforms.py:
```diff
@@ -282,20 +281,494 @@ def index_range(stack):
+            builder.end_list()
+        return builder
+
+    @numba.jit
```
scope issue here as well.

lgray commented on src/coffea/nanoevents/transforms.py:
```diff
@@ -282,20 +281,494 @@ def index_range(stack):
+
+
+def get_array_from_indices_xyzrecord_target(indices, target):
+    @numba.jit
```
another local scope + numba issue

prayagyadav committed:
Fixed Scope Issues

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

lgray committed:
Merge branch 'master' into edm4hep-schema

lgray committed:
Merge branch 'master' into edm4hep-schema

lgray committed:
Merge branch 'master' into edm4hep-schema

lgray committed:
Merge branch 'master' into edm4hep-schema

prayagyadav committed:
Typenames could be added to branch forms in delayed=False mode

prayagyadav committed:
eager and delayed typenames work given a change in uproot5

prayagyadav committed:
All the schemas work with the typename change; fullsim samples are readable with the EDM4HEPSchema_v00_10_05

prayagyadav committed:
Merge branch 'scikit-hep:master' into test-typenames

prayagyadav committed:
w

prayagyadav committed:
Everything works

prayagyadav committed:
cleaned

prayagyadav committed:
Merge branch 'scikit-hep:master' into test-typenames

prayagyadav committed:
support for extend ak_add_doc

prayagyadav committed:
Merge branch 'scikit-hep:master' into edm4hep-schema

prayagyadav committed:
bring back the uproot dependency

prayagyadav committed:
Adding the prospective functionality to recognize any edm4hep style root file
(Don't merge yet: This needs the new feature:'infer typename in uproot.dask' to be merged for this to work)

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

prayagyadav committed:
fixed spelling mistake

prayagyadav committed:
Merge branch 'edm4hep-schema' of https://github.com/prayagyadav/coffea into edm4hep-schema

lgray committed:
Merge branch 'master' into edm4hep-schema

prayagyadav committed:
cleaned schemas/edm4hep.py

prayagyadav committed:
cleaned all the other files

prayagyadav committed:
Many Tests added

prayagyadav committed:
Merge branch 'master' into edm4hep-schema

prayagyadav committed:
python ver < 3.10 compatibility

prayagyadav committed:
precommit

prayagyadav committed:
Added typename in dataset_tools, test_dataset_tools.py and test_nanoevents_physlite.py

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

prayagyadav commented on src/coffea/nanoevents/assets/__init__.py:
```diff
@@ -0,0 +1,24 @@
+
+import yaml
+
+root_dir = os.path.dirname(os.path.abspath(__file__))
```
aah , this section needs improvement, I guess. Maybe there is a way to load only the yaml version that I need at runtime, instead of loading every version.

prayagyadav replied:
@lgray I would like to fix this before merge.

lgray replied:
Sure, you can keep a dictionary of versions to functions that load the info just in time, instead of pre-loading everything like you have here. That should achieve what you want.

I am also somewhat unsure if we want to have non-python assets be tied into the module structure, may warrant a small reorganization.

lgray replied:
something like:
```python3
versions  = [
    # as they are now
]

def _load_edm4hep_version(yamlfile):
    with open(yamlfile) as f:
         return yaml.safe_load(f)

edm4hep_ver  = {
    k: partial(
        _load_edm4hep_version,
        yamlfile="/".join([root_dir, f"edm4hep_v{k}.yaml"]))
    ) for k in versions
}
```
Then later in calling code: `edm4hep_ver[version]()`

prayagyadav replied:
aah, thats smart. Will implement it.

lgray replied:
You should also use `os.path.join(root_dir, f"edm4hep_v{k}.yaml")`rather than `"/".join(...)`

nsmith- replied:
The safest way to locate the data would be with `importlib.resources.files("coffea.nanoevents.assets")` (see e.g. https://github.com/cms-nanoAOD/correctionlib/blob/master/src/correctionlib/util.py)

prayagyadav committed:
lazy load yaml

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

lgray commented on src/coffea/nanoevents/schemas/edm4hep.py:
```diff
@@ -193,6 +193,7 @@ def version(cls, ver="latest"):
             raise ValueError(
                 "The given version {ver} is not found. Available versions are : 00.99.01,00.99.00,00.10.05,00.10.04,00.10.03,00.10.02 and 00.10.01"
```
The version list can be a `','.join(...)` on the keys of the version dict.

lgray committed:
Merge branch 'master' into edm4hep-schema

prayagyadav committed:
better imports; fixed hard-coded raise error in EDM4HEPSchema.version

lgray committed:
edm4hep reading requires uproot 5.6.0

prayagyadav committed:
Merge branch 'master' into edm4hep-schema
